core/src/main/scala/kafka/cluster/Replica.scala

```diff
@@ -165,9 +168,16 @@ class Replica(val brokerId: Int,
       s"non-local replica $brokerId"))
   }
 
-  def convertHWToLocalOffsetMetadata() = {
+  /*
+   * Convert hw to local offset metadata by reading the log at the hw offset.
+   * If the hw offset is out of range, return the first offset of the first log segment as the offset metadata.
+   */
+  def convertHWToLocalOffsetMetadata() {
     if (isLocal) {
-      highWatermarkMetadata = log.get.convertToOffsetMetadata(highWatermarkMetadata.messageOffset)
+      highWatermarkMetadata = log.get.convertToOffsetMetadata(highWatermarkMetadata.messageOffset).getOrElse {
+        val firstOffset = log.get.logSegments.head.baseOffset
```

Comment:
Should we use log.logStartOffset for firstOffset?

Reply:
@junrao Great point. Yeah logStartOffset should be better.

Reply:
@lindong28 Were you planning to submit a follow-up for this change?

Reply:
@hachikuji Sure. I opened https://github.com/apache/kafka/pull/4722 for this change.